### DISCUSSION TITLE
Consider Apache Arrow for data structures and data exchange

### DISCUSSION THREAD

**Discussion by michitux** [networkit/networkit#677] [Ideas]
Currently, we use standard C++ vectors in many places for storing both the graphs itself and results from algorithms. When executing algorithms from Python, results are frequently converted into Python lists (e.g., centrality scores), effectively copying all data. [Apache Arrow](https://arrow.apache.org/) "defines a language-independent columnar memory format for flat and hierarchical data". Apache Arrow also provides a C++ library. At the same time, there are Python and [Cython bindings](https://arrow.apache.org/docs/python/extending.html#cython-api) which should allow making the data structures that have been created in C++ available in Python without copying any data. Under certain conditions (e.g., just a series of integer or floating point values) these data structures can also be [converted to Pandas](https://arrow.apache.org/docs/python/pandas.html) objects [without any copying](https://arrow.apache.org/docs/python/pandas.html#zero-copy-series-conversions). The C++ API also implements I/O function for reading [Apache Parquet](https://arrow.apache.org/docs/cpp/parquet.html) files which might be interesting as a space-efficient storage format.

I think adopting Apache Arrow for all non-temporary data structures in NetworKit would be a great improvement in terms of performance, memory usage and interoperability in the data science ecosystem.

This is of course also related to #655 (node and edge metadata) and #420 (new graph data structure).

**Comment by avdgrinten** [comment]
I don't think that we should rely on a specific data format for `Graph`s (even though the current storage strategy is not as efficient as it could be). However, this (or a similar strategy that allows zero-copy sharing with Python code) might be useful for other data types, e.g., data types that are returned from algorithms.
